examples/course_manual_publisher/lib/course_manual_publisher.rb

```diff
@@ -0,0 +1,269 @@
+          "Content-MD5" => checksum_for(full_path),
+          "Content-Length" => File.size(full_path),
+          "x-ms-blob-type" => 'BlockBlob'
+        }
+      )
+    end
+  end
+
+  def upload_content
+    request(
+      method: :put,
+      payload: {
+        status: "content_complete",
+        content: processed_html,
+        attachment_upload_data: initial_attachment_upload_data
+      }
+    )
+  end
+
+  def poll_for_state(desired_states, failure_states)
+    desired_states, failure_states =
+      Array.wrap(desired_states), Array.wrap(failure_states)
+
+    desired_states_str = desired_states.map(&:humanize).map(&:downcase).join(' or ')
+
+    1.upto(MAX_POLLS) do
+      current_state = request.try(:[], 'state') ||
+                        raise(UnexpectedStateError, "No state found")
+
+      current_state_str = current_state.humanize.downcase
+
+      if desired_states.include?(current_state)
+        puts "Now #{current_state_str}"
+        return current_state
+      elsif failure_states.include?(current_state)
+        raise UnexpectedStateError, "Failure: unexpectedly #{current_state_str}"
+      else
+        puts "Currently #{current_state_str}, waiting until #{desired_states_str}"
+        sleep POLL_SLEEP_SECS
+      end
+    end
+
+    raise UnexpectedStateError, "Timed out waiting until #{desired_states_str}"
+  end
+
+  def request(method: :get, payload: nil)
+    r = RestClient::Request.execute(
+      method:,
+      url:,
+      payload: payload&.to_json,
+      headers: { 
+        "Accept" => "application/json",
+        "Content-Type" => "application/json",
+        "Authorization" => authorization
+      }
+    )
+
+    JSON.parse(r.body) if r.body.present?
+  end
+
+  def raw_html
+    File.read(html_file)
+  end
+
+  def url
+    "https://#{app_hostname}/api/v1/courses/#{course_id}/course_manual/versions/draft"
+  end
+
+  def authorization
+    "Basic " + Base64.strict_encode64("#{api_key}:#{api_secret}")
+  end
+
+  def checksum_for(file)
+    Base64.strict_encode64(
+      Digest::MD5.hexdigest(
+        File.read(file)
+      ).scan(/../).map { |x| x.hex.chr }.join
+    )
+  end
+
+  def content_type_for(file)
+    MIME::Types.type_for(file).first&.content_type || 
+      raise("Can't find content-type for #{f}")
+  end
+
+  # Paths are treated as local attachments to be uploaded / relinked in HTML if
+  # a) URI format with scheme of 'file' => absolute local path; OR
+  # b) non-URI format, doesn't start with / => relative local path (to HTML)
+  def resolve_path_if_local(path)
+    return nil if path.nil?
+
+    res_path = path.gsub("\\","/")
+
+    if res_path.include?(":")
+      if res_path.split(":").first == "file"
+        File.expand_path(res_path.gsub(/^file\:\/\/.*?\//, "/") )
+      end
+    elsif !res_path.start_with?("/")
+      File.expand_path(File.join(base_dir, res_path))
+    end
+  end
+
+  def die(msg)
+    STDERR.puts msg
+    exit 1
+  end
```

Comment:
This method is the same as Ruby's native `abort` https://ruby-doc.org/3.2.0/Kernel.html#method-i-abort.